### PR TITLE
refactor: goclient installer

### PR DESCRIPTION
**Pull request by hxcGit**
Signed-off-by: hxcGit <houxc_mail@163.com>

## Pre-Checklist

Note: please complete **_ALL_** items in the following checklist.

- [ ] I have read through the [CONTRIBUTING.md](https://github.com/devstream-io/devstream/blob/main/CONTRIBUTING.md) documentation.
- [ ] My code has the necessary comments and documentation (if needed).
- [ ] I have added relevant tests

## Description

- Refactor goclient to plugininstall way
- `zentao` plugin has been tested
## Related Issues

#850 
## New Behavior (screenshots if needed)

```
{22-07-22 21:55}hxc-macbook-air:devstream@refactor-go-client✗✗✗✗✗✗ hxc% ./dtm apply -f zentao.yaml -y --debug 
2022-07-22 21:55:59 ℹ [INFO]  Log level is: debug.
2022-07-22 21:55:59 ℹ [INFO]  Apply started.
2022-07-22 21:55:59 λ [DEBUG]  Original config: 
---
# core config
varFile: ''
toolFile: ''
state: # state config, backend can be local or s3
  backend: local
  options:
    stateFile: devstream.state

---
# plugins config
tools:
  # name of the tool
  - name: zentao
    # id of the tool instance
    instanceID: default
    # format: name.instanceID; If specified, dtm will make sure the dependency is applied first before handling this tool
    dependsOn: []
    # options for the plugin
    options:
      # namespace for ZenTao application
      namespace: 'zentao'
      # storageClassName used to match pv and pvc
      storageClassName: 'zentao-storage'
      # two PersistentVolumes for ZenTao and mysql should be specified
      persistentVolumes:
          # name of ZenTao pv
        - pvName: "zentao-pv"
          # capacity of ZenTao pv
          pvCapacity: "1G"
          # name of mysql pv
        - pvName: "mysql-pv"
          # capacity of mysql pv
          pvCapacity: "1G"
      # two PersistentVolumeClaims for ZenTao and mysql should be specified
      persistentVolumeClaims:
          # name of ZenTao pvc
        - pvcName: "zentao-pvc"
          # capacity of ZenTao pvc
          pvcCapacity: "1G"
          # name of mysql pvc
        - pvcName: "mysql-pvc"
          # capacity of mysql pvc
          pvcCapacity: "1G"
      # ZenTao application is deployed by K8s Deployment
      deployment:
        # name of ZenTao deployment
        name: 'zentao-dp'
        # number of application replica
        replicas: 1
        # ZenTao image
        image: 'easysoft/zentao:latest'
        env:
          key: 'MYSQL_ROOT_PASSWORD'
          # initial password value for mysql database, you can specify any value you like
          value: '12345678'
      # ZenTao application is exposed via K8s Service
      service:
        # name of ZenTao service
        name: 'zentao-svc'
        # nodePort of ZenTao service, currently ZenTao plugin only support `nodePort` type
        nodePort: 30081


2022-07-22 21:55:59 ℹ [INFO]  Got Backend from config: local
2022-07-22 21:55:59 ℹ [INFO]  Using dir <.devstream> to store plugins.
2022-07-22 21:56:00 λ [DEBUG]  The global manager m is not initialized.
2022-07-22 21:56:00 ℹ [INFO]  Using local backend. State file: devstream.state.
2022-07-22 21:56:00 λ [DEBUG]  Used the Backend: local.
2022-07-22 21:56:00 λ [DEBUG]  isForce:false
2022-07-22 21:56:00 λ [DEBUG]  TOOL zentao.default dependency already solved

2022-07-22 21:56:00 λ [DEBUG]  BATCH: [{zentao default [] map[deployment:map[env:map[key:MYSQL_ROOT_PASSWORD value:12345678] image:easysoft/zentao:latest name:zentao-dp replicas:1] namespace:zentao persistentVolumeClaims:[map[pvcCapacity:1G pvcName:zentao-pvc] map[pvcCapacity:1G pvcName:mysql-pvc]] persistentVolumes:[map[pvCapacity:1G pvName:zentao-pv] map[pvCapacity:1G pvName:mysql-pv]] service:map[name:zentao-svc nodePort:30081] storageClassName:zentao-storage]}]
2022-07-22 21:56:00 λ [DEBUG]  Changes for the plan:
2022-07-22 21:56:00 λ [DEBUG]  Change - 1/1 -> 
{
  ActionName: Create,
  Tool: {Name: zentao, InstanceID: default}}
}
2022-07-22 21:56:00 ℹ [INFO]  Tool (zentao/default) found in config but doesn't exist in the state, will be created.
2022-07-22 21:56:00 ℹ [INFO]  Start executing the plan.
2022-07-22 21:56:00 ℹ [INFO]  Changes count: 1.
2022-07-22 21:56:00 λ [DEBUG]  TOOL zentao.default dependency already solved

2022-07-22 21:56:00 λ [DEBUG]  BATCH: [{zentao default [] map[deployment:map[env:map[key:MYSQL_ROOT_PASSWORD value:12345678] image:easysoft/zentao:latest name:zentao-dp replicas:1] namespace:zentao persistentVolumeClaims:[map[pvcCapacity:1G pvcName:zentao-pvc] map[pvcCapacity:1G pvcName:mysql-pvc]] persistentVolumes:[map[pvCapacity:1G pvName:zentao-pv] map[pvCapacity:1G pvName:mysql-pv]] service:map[name:zentao-svc nodePort:30081] storageClassName:zentao-storage]}]
2022-07-22 21:56:00 ℹ [INFO]  -------------------- [  Processing progress: 1/1.  ] --------------------
2022-07-22 21:56:00 ℹ [INFO]  Processing: (zentao/default) -> Create ...
2022-07-22 21:56:00 λ [DEBUG]  Tool's raw changes are: map[deployment:map[env:map[key:MYSQL_ROOT_PASSWORD value:12345678] image:easysoft/zentao:latest name:zentao-dp replicas:1] namespace:zentao persistentVolumeClaims:[map[pvcCapacity:1G pvcName:zentao-pvc] map[pvcCapacity:1G pvcName:mysql-pvc]] persistentVolumes:[map[pvCapacity:1G pvName:zentao-pv] map[pvCapacity:1G pvName:mysql-pv]] service:map[name:zentao-svc nodePort:30081] storageClassName:zentao-storage].
2022-07-22 21:56:00 λ [DEBUG]  Start Execute PreInstall Operations...
2022-07-22 21:56:00 λ [DEBUG]  Start Execute Install Operations...
2022-07-22 21:56:00 λ [DEBUG]  Prepare to create the namespace: zentao.
2022-07-22 21:56:00 λ [DEBUG]  Failed to get the kubecondig from env. Prepare to get it from home dir.
2022-07-22 21:56:00 λ [DEBUG]  The namespace zentao has been created.
2022-07-22 21:56:00 λ [DEBUG]  Failed to get the kubecondig from env. Prepare to get it from home dir.
2022-07-22 21:56:00 λ [DEBUG]  Prepare to create persistentVolume: zentao-pv.
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolume < zentao-pv > has created.
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolume zentao-pv has been created.
2022-07-22 21:56:00 λ [DEBUG]  Prepare to create persistentVolume: mysql-pv.
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolume < mysql-pv > has created.
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolume mysql-pv has been created.
2022-07-22 21:56:00 λ [DEBUG]  Failed to get the kubecondig from env. Prepare to get it from home dir.
2022-07-22 21:56:00 λ [DEBUG]  Prepare to create persistentVolumeClaim: zentao-pvc.
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolumeCliam option for creation is &PersistentVolumeClaim{ObjectMeta:{zentao-pvc      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteOnce],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1 9} {<nil>} 1G DecimalSI},},},VolumeName:,Selector:nil,StorageClassName:*zentao-storage,VolumeMode:nil,DataSource:nil,DataSourceRef:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},},}
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolumeClaim < zentao-pvc > has created.
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolumeClaim zentao-pvc has been created.
2022-07-22 21:56:00 λ [DEBUG]  Prepare to create persistentVolumeClaim: mysql-pvc.
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolumeCliam option for creation is &PersistentVolumeClaim{ObjectMeta:{mysql-pvc      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteOnce],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1 9} {<nil>} 1G DecimalSI},},},VolumeName:,Selector:nil,StorageClassName:*zentao-storage,VolumeMode:nil,DataSource:nil,DataSourceRef:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},},}
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolumeClaim < mysql-pvc > has created.
2022-07-22 21:56:00 λ [DEBUG]  The PersistentVolumeClaim mysql-pvc has been created.
2022-07-22 21:56:00 λ [DEBUG]  Failed to get the kubecondig from env. Prepare to get it from home dir.
2022-07-22 21:56:00 λ [DEBUG]  The Deployment zentao-dp has been created.
2022-07-22 21:56:00 λ [DEBUG]  The Deployment zentao-dp has been created.
2022-07-22 21:56:00 λ [DEBUG]  Failed to get the kubecondig from env. Prepare to get it from home dir.
2022-07-22 21:56:00 λ [DEBUG]  The Service zentao-svc has been created.
2022-07-22 21:56:00 λ [DEBUG]  The Service zentao-svc has been created.
2022-07-22 21:56:00 λ [DEBUG]  Failed to get the kubecondig from env. Prepare to get it from home dir.
2022-07-22 21:56:05 λ [DEBUG]  Retry check deployment status 0 times
2022-07-22 21:56:05 ℹ [INFO]  The deployment zentao-dp is ready.
2022-07-22 21:56:05 λ [DEBUG]  Start Execute Status Operations...
2022-07-22 21:56:05 λ [DEBUG]  Failed to get the kubecondig from env. Prepare to get it from home dir.
2022-07-22 21:56:05 λ [DEBUG]  The deployment zentao-dp is ready.
2022-07-22 21:56:05 λ [DEBUG]  The service zentao-svc is ready.
2022-07-22 21:56:05 λ [DEBUG]  Start -> StatesMap now is:

2022-07-22 21:56:05 ✔ [SUCCESS]  Tool (zentao/default) Create done.
2022-07-22 21:56:05 λ [DEBUG]  End -> StatesMap now is:
zentao_default:
  name: zentao
  instanceid: default
  dependson: []
  options:
    deployment:
      env:
        key: MYSQL_ROOT_PASSWORD
        value: "12345678"
      image: easysoft/zentao:latest
      name: zentao-dp
      replicas: 1
    namespace: zentao
    persistentVolumeClaims:
      - pvcCapacity: 1G
        pvcName: zentao-pvc
      - pvcCapacity: 1G
        pvcName: mysql-pvc
    persistentVolumes:
      - pvCapacity: 1G
        pvName: zentao-pv
      - pvCapacity: 1G
        pvName: mysql-pv
    service:
      name: zentao-svc
      nodePort: 30081
    storageClassName: zentao-storage
  resource:
    running: true

2022-07-22 21:56:05 ℹ [INFO]  -------------------- [  Processing done.  ] --------------------
2022-07-22 21:56:05 ✔ [SUCCESS]  All plugins applied successfully.
2022-07-22 21:56:05 ✔ [SUCCESS]  Apply finished.
{22-07-22 21:56}hxc-macbook-air:devstream@refactor-go-client✗✗✗✗✗✗ hxc% ./dtm delete -f zentao.yaml -y --debug
2022-07-22 21:56:12 ℹ [INFO]  Log level is: debug.
2022-07-22 21:56:12 ℹ [INFO]  Delete started.
2022-07-22 21:56:12 λ [DEBUG]  Original config: 
---
# core config
varFile: ''
toolFile: ''
state: # state config, backend can be local or s3
  backend: local
  options:
    stateFile: devstream.state

---
# plugins config
tools:
  # name of the tool
  - name: zentao
    # id of the tool instance
    instanceID: default
    # format: name.instanceID; If specified, dtm will make sure the dependency is applied first before handling this tool
    dependsOn: []
    # options for the plugin
    options:
      # namespace for ZenTao application
      namespace: 'zentao'
      # storageClassName used to match pv and pvc
      storageClassName: 'zentao-storage'
      # two PersistentVolumes for ZenTao and mysql should be specified
      persistentVolumes:
          # name of ZenTao pv
        - pvName: "zentao-pv"
          # capacity of ZenTao pv
          pvCapacity: "1G"
          # name of mysql pv
        - pvName: "mysql-pv"
          # capacity of mysql pv
          pvCapacity: "1G"
      # two PersistentVolumeClaims for ZenTao and mysql should be specified
      persistentVolumeClaims:
          # name of ZenTao pvc
        - pvcName: "zentao-pvc"
          # capacity of ZenTao pvc
          pvcCapacity: "1G"
          # name of mysql pvc
        - pvcName: "mysql-pvc"
          # capacity of mysql pvc
          pvcCapacity: "1G"
      # ZenTao application is deployed by K8s Deployment
      deployment:
        # name of ZenTao deployment
        name: 'zentao-dp'
        # number of application replica
        replicas: 1
        # ZenTao image
        image: 'easysoft/zentao:latest'
        env:
          key: 'MYSQL_ROOT_PASSWORD'
          # initial password value for mysql database, you can specify any value you like
          value: '12345678'
      # ZenTao application is exposed via K8s Service
      service:
        # name of ZenTao service
        name: 'zentao-svc'
        # nodePort of ZenTao service, currently ZenTao plugin only support `nodePort` type
        nodePort: 30081


2022-07-22 21:56:12 ℹ [INFO]  Got Backend from config: local
2022-07-22 21:56:12 ℹ [INFO]  Using dir <.devstream> to store plugins.
2022-07-22 21:56:12 λ [DEBUG]  The global manager m is not initialized.
2022-07-22 21:56:12 ℹ [INFO]  Using local backend. State file: devstream.state.
2022-07-22 21:56:12 λ [DEBUG]  Used the Backend: local.
2022-07-22 21:56:12 λ [DEBUG]  Got a state from the backend: zentao_default -> {zentao default [] map[deployment:map[env:map[key:MYSQL_ROOT_PASSWORD value:12345678] image:easysoft/zentao:latest name:zentao-dp replicas:1] namespace:zentao persistentVolumeClaims:[map[pvcCapacity:1G pvcName:zentao-pvc] map[pvcCapacity:1G pvcName:mysql-pvc]] persistentVolumes:[map[pvCapacity:1G pvName:zentao-pv] map[pvCapacity:1G pvName:mysql-pv]] service:map[name:zentao-svc nodePort:30081] storageClassName:zentao-storage] map[running:true]}.
2022-07-22 21:56:12 λ [DEBUG]  isForce:false
2022-07-22 21:56:12 λ [DEBUG]  TOOL zentao.default dependency already solved

2022-07-22 21:56:12 λ [DEBUG]  BATCH: [{zentao default [] map[deployment:map[env:map[key:MYSQL_ROOT_PASSWORD value:12345678] image:easysoft/zentao:latest name:zentao-dp replicas:1] namespace:zentao persistentVolumeClaims:[map[pvcCapacity:1G pvcName:zentao-pvc] map[pvcCapacity:1G pvcName:mysql-pvc]] persistentVolumes:[map[pvCapacity:1G pvName:zentao-pv] map[pvCapacity:1G pvName:mysql-pv]] service:map[name:zentao-svc nodePort:30081] storageClassName:zentao-storage]}]
2022-07-22 21:56:12 λ [DEBUG]  Changes for the plan:
2022-07-22 21:56:12 λ [DEBUG]  Change - 1/1 -> 
{
  ActionName: Delete,
  Tool: {Name: zentao, InstanceID: default}}
}
2022-07-22 21:56:12 ℹ [INFO]  Tool (zentao/default) will be deleted.
2022-07-22 21:56:12 ℹ [INFO]  Start executing the plan.
2022-07-22 21:56:12 ℹ [INFO]  Changes count: 1.
2022-07-22 21:56:12 λ [DEBUG]  TOOL zentao.default dependency already solved

2022-07-22 21:56:12 λ [DEBUG]  BATCH: [{zentao default [] map[deployment:map[env:map[key:MYSQL_ROOT_PASSWORD value:12345678] image:easysoft/zentao:latest name:zentao-dp replicas:1] namespace:zentao persistentVolumeClaims:[map[pvcCapacity:1G pvcName:zentao-pvc] map[pvcCapacity:1G pvcName:mysql-pvc]] persistentVolumes:[map[pvCapacity:1G pvName:zentao-pv] map[pvCapacity:1G pvName:mysql-pv]] service:map[name:zentao-svc nodePort:30081] storageClassName:zentao-storage]}]
2022-07-22 21:56:12 ℹ [INFO]  -------------------- [  Processing progress: 1/1.  ] --------------------
2022-07-22 21:56:12 ℹ [INFO]  Processing: (zentao/default) -> Delete ...
2022-07-22 21:56:12 λ [DEBUG]  Tool's raw changes are: map[deployment:map[env:map[key:MYSQL_ROOT_PASSWORD value:12345678] image:easysoft/zentao:latest name:zentao-dp replicas:1] namespace:zentao persistentVolumeClaims:[map[pvcCapacity:1G pvcName:zentao-pvc] map[pvcCapacity:1G pvcName:mysql-pvc]] persistentVolumes:[map[pvCapacity:1G pvName:zentao-pv] map[pvCapacity:1G pvName:mysql-pv]] service:map[name:zentao-svc nodePort:30081] storageClassName:zentao-storage].
2022-07-22 21:56:12 λ [DEBUG]  Start Execute PreInstall Operations...
2022-07-22 21:56:12 λ [DEBUG]  Start Execute Install Operations...
2022-07-22 21:56:12 λ [DEBUG]  Failed to get the kubecondig from env. Prepare to get it from home dir.
2022-07-22 21:56:12 λ [DEBUG]  Prepare to delete PVC.
2022-07-22 21:56:12 λ [DEBUG]  Prepare to delete PV.
2022-07-22 21:56:12 λ [DEBUG]  Prepare to delete namespace.
2022-07-22 21:56:13 λ [DEBUG]  Start -> StatesMap now is:
zentao_default:
  name: zentao
  instanceid: default
  dependson: []
  options:
    deployment:
      env:
        key: MYSQL_ROOT_PASSWORD
        value: "12345678"
      image: easysoft/zentao:latest
      name: zentao-dp
      replicas: 1
    namespace: zentao
    persistentVolumeClaims:
      - pvcCapacity: 1G
        pvcName: zentao-pvc
      - pvcCapacity: 1G
        pvcName: mysql-pvc
    persistentVolumes:
      - pvCapacity: 1G
        pvName: zentao-pv
      - pvCapacity: 1G
        pvName: mysql-pv
    service:
      name: zentao-svc
      nodePort: 30081
    storageClassName: zentao-storage
  resource:
    running: true

2022-07-22 21:56:13 ℹ [INFO]  Prepare to delete 'zentao_default' from States.
2022-07-22 21:56:13 ✔ [SUCCESS]  Tool (zentao/default) delete done.
2022-07-22 21:56:13 λ [DEBUG]  End -> StatesMap now is:

2022-07-22 21:56:13 ℹ [INFO]  -------------------- [  Processing done.  ] --------------------
2022-07-22 21:56:13 ✔ [SUCCESS]  All plugins deleted successfully.
2022-07-22 21:56:13 ✔ [SUCCESS]  Delete finished.
```

### CODE REVIEW

File: internal/pkg/plugininstaller/goclient/option.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/devstream-io/devstream/internal/pkg/plugininstaller"
+)
+
+// Options is the struct for parameters used by the goclient install config.
+type Options struct {
+	Namespace              string                  `validate:"required"`
+	StorageClassName       string                  `validate:"required"`
+	PersistentVolumes      []PersistentVolume      `validate:"required"`
+	PersistentVolumeClaims []PersistentVolumeClaim `validate:"required"`
+	Deployment             Deployment              `validate:"required"`
```
Comment (iyear) [inline internal/pkg/plugininstaller/goclient/option.go:15]: `required` does not make sense for non-pointer structs. Remove `required` or change to `*Deployment`

File: internal/pkg/plugininstaller/goclient/utils.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+// Check whether deployment is ready and service exists
+func checkDeploymentsAndServicesReady(kubeClient *k8s.Client, opts *Options) (bool, error) {
+	namespace := opts.Namespace
+	deploy := opts.Deployment.Name
+	svc := opts.Service.Name
+
+	dp, err := kubeClient.GetDeployment(namespace, deploy)
+	if err != nil {
+		log.Debugf("Get deployment err: %s", err.Error())
+		if strings.Contains(err.Error(), "not found") {
```
Comment (iyear) [inline internal/pkg/plugininstaller/goclient/utils.go:21]: Is it possible to use `kerr.IsNotFound(err)` here? I don't know😂

File: internal/pkg/plugininstaller/goclient/option.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+type PersistentVolumeClaim struct {
+	PVCName     string `validate:"required"`
+	PVCCapacity string `validate:"required"`
+}
+
+type Deployment struct {
+	Name     string `validate:"required"`
+	Replicas int    `validate:"required"`
+	Image    string `validate:"required"`
+	Env      Env    `validate:"required"`
```
Comment (iyear) [inline internal/pkg/plugininstaller/goclient/option.go:34]: Is env the only one that can be set?

File: internal/pkg/plugininstaller/goclient/utils.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+			return false, nil
+		}
+		return false, err
+	}
+	log.Debugf("The service %s is ready.", svc)
+
+	return true, nil
+}
+
+func DeleteService(kubeClient *k8s.Client, opts *Options) error {
```
[the remark this answers is not in the document]
Comment (hxcGit) [reply internal/pkg/plugininstaller/goclient/utils.go]: Done

File: internal/pkg/plugininstaller/goclient/utils.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+func getDevstreamNamespace(kubeClient *k8s.Client) (*corev1.NamespaceList, error) {
+	return kubeClient.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{LabelSelector: "created_by=DevStream"})
+}
+
+// Check whether the given namespace is created by dtm
+// If the given namespace has label "created_by=DevStream", we'll control it.
+// 1. The specified namespace is created by zentao plugin, then it should be deleted
+//    when errors are encountered during creation or `dtm delete`.
+// 2. The specified namespace is controlled by user, maybe they want to deploy zentao in
+//    an existing namespace or other situations, then we should not delete this namespace.
```
[the remark this answers is not in the document]
Comment (hxcGit) [reply internal/pkg/plugininstaller/goclient/utils.go:70]: Done

File: internal/pkg/plugininstaller/goclient/installer.go [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+		}
+		log.Debugf("The Service %s has been created.", svc.Name)
+
+		return nil
+	}
+}
+
+// Check application status by goclient with retry times
+func WaitForReady(retry int) plugininstaller.BaseOperation {
+	return func(options plugininstaller.RawOptions) error {
```
[the remark this answers is not in the document]
Comment (hxcGit) [reply internal/pkg/plugininstaller/goclient/installer.go:292]: Done

File: internal/pkg/plugininstaller/goclient/option.go (hunk 3 above)
Comment (hxcGit) [reply internal/pkg/plugininstaller/goclient/option.go]: []Env is better, I'll fix this.

**Review by iyear — APPROVED** [review]

**Comment by iyear** [comment]
Nice refactoring!

File: internal/pkg/plugininstaller/goclient/utils.go [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+	return nil
+}
+
+// Check whether the given namespace is created by dtm
+// If the given namespace has label "created_by=DevStream", we'll control it.
+// 1. The specified namespace is created by zentao plugin, then it should be deleted
+//    when errors are encountered during creation or `dtm delete`.
+// 2. The specified namespace is controlled by user, maybe they want to deploy zentao in
+//    an existing namespace or other situations, then we should not delete this namespace.
+func isDevstreamNSExists(kubeClient *k8s.Client, namespace string) (bool, error) {
```
Comment (aFlyBird0) [inline internal/pkg/plugininstaller/goclient/utils.go:70]: You should export it so that other plugins could call this function.

It's a useful function.👏

Comment (hxcGit) [reply internal/pkg/plugininstaller/goclient/utils.go]: If we want use this function outside `plugininstall/goclient`. I think it's better to put it in `pkg/util/k8s`.

**Review by IronCore864 — APPROVED** [review]

**Pull request merged**